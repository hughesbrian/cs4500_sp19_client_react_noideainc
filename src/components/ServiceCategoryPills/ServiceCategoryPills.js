import React from 'react'

const ServiceCategoryPills = ({serviceCategories, updateScore}) => {
    const listServiceCategory = serviceCategories.map( (item, index) => {
        let icon = <i className="fa fa-bookmark"></i>
        let iconRule = {
            "Pets": <i className="fa fa-paw"/>,
            "Wellness": <i className="fa fa-heart"/>,
            "Business": <i className="fa fa-briefcase"/>, 
            "Events": <i className="fa fa-calendar"></i>
        }
        if (item.title in iconRule) {
            icon = iconRule[item.title]
        }
        return (
            <li onClick={() => updateScore(item)} className="nav-item" key={index}>
                <a className="nav-link btn-lg text-center" href="/services">
                    {icon}
                    <br/>
                    {item.title}
                </a>
            </li>
        )
    } )

    return (<ul className="nav nav-pills nav-fill wd-shadow wd-padding-20">
        <li className="nav-item">
            <a className="nav-link btn-lg text-center"
               href="/services">
                <i className="fa fa-home"/>
                <br/>
                Home
            </a>
        </li>
        {listServiceCategory}
        <li className="nav-item">
            <a className="nav-link btn-lg text-center"
               href="/services">
                <i className="fa fa-ellipsis-h"/>
                <br/>
                More
            </a>
        </li>
    </ul>)
}

export default ServiceCategoryPills