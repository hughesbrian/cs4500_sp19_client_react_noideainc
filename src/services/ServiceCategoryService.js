const baseURL = "https://cs4500-sp19-noideainc.herokuapp.com";
//const baseURL = "http://localhost:8080";

export default class ServiceCategoryService {
    static instance = null;
    static getInstance() {
        if (ServiceCategoryService.instance === null) {
            ServiceCategoryService.instance = new ServiceCategoryService()
        }
        return this.instance
    }

    findServiceCategoryById = categoryId =>
        fetch(baseURL + `/api/categories/${categoryId}`)
            .then(response => response.json())

    findAllServiceCategories = () =>
        fetch(baseURL +  `/api/categories`)
            .then(response => response.json())

    pagedServiceCategories = (page, count) =>
        fetch(baseURL + `/api/categories/paged?page=${page}&count=${count}`)
            .then(response => response.json())

    createServiceCategory = sc => {
        delete sc.id;
        return fetch(baseURL + `/api/categories`, {
            method: 'post',
            body: JSON.stringify(sc),
            headers: {
                'content-type': 'application/json'
            }
        })
        .then(response => response.json())
    }

    updateServiceCategory = sc =>
        fetch(baseURL + `/api/categories/${sc.id}`, {
            method: 'put',
            body: JSON.stringify(sc),
            headers: {
                'content-type': 'application/json'
            }
        }).then(response => response.json())
    
    updateServiceCategoryScore = sc =>
        fetch(baseURL + `/api/categories/score/${sc.id}`, {
            method: 'put',
            body: JSON.stringify(sc),
            headers: {
                'content-type': 'application/json'
            }
        }).then(response => response.json())

    deleteServiceCategory = id =>
        fetch(baseURL + `/api/categories/${id}`, {
            method: 'delete'
        })

    findAllServicesByCategoryName = catName =>
        fetch(baseURL +  `/api/${catName}`)
            .then(response => response.json())
}
