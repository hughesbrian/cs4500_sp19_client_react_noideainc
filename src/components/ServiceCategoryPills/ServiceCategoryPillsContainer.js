import React from 'react'
import ServiceCategoryPills from './ServiceCategoryPills'
import ServiceCategoryService from '../../services/ServiceCategoryService'

class ServiceCategoryPillsContainer extends React.Component {
    constructor(props) {
        super(props);
        this.serviceCategoryService = ServiceCategoryService.getInstance();
        this.state = {
            serviceCategories: []
        }
    }

    componentDidMount() {
        this.findAllServiceCategories()
    }

    findAllServiceCategories = () =>
        this.serviceCategoryService
            .findAllServiceCategories()
            .then(serviceCategories =>
                this.setState({
                    serviceCategories: this.findTopPopularity(serviceCategories)
                })
            )

    // find the top three popularity service categories and display them in pills
    findTopPopularity = (serviceCategories) => {
        let filterArray = [];
        serviceCategories.forEach((item) => {
            let start = 0;
            while (filterArray.length > start) {
                if (item.score > filterArray[start]) {
                    break;
                }
                start++;
            }
            filterArray.splice(start, 1, item)
            if (filterArray.length > 3) {
                filterArray = filterArray.slice(0, 3);
            }
        })
        console.log(filterArray)
        return filterArray;
    }

    // increase score when one service category icon is clicked
    updateScore = (serviceCategory) => {
        serviceCategory.score = serviceCategory.score + 1;
        //alert("update score")
        //alert(serviceCategory.score)
        this.serviceCategoryService
            .updateServiceCategoryScore(serviceCategory)
            .then()
    }

    render = () =>
        <ServiceCategoryPills serviceCategories={this.state.serviceCategories} updateScore={this.updateScore} />

}

export default ServiceCategoryPillsContainer
