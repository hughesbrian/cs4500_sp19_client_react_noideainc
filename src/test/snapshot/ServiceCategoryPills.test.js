import React from 'react'
import ServiceCategoryPills from '../../components/ServiceCategoryPills/ServiceCategoryPills'
import TestRenderer from 'react-test-renderer'
import serviceCategories from '../MockData/ServiceCategories2.mock.json'
import ServiceCategoryService from '../../services/ServiceCategoryService'
const serviceCategoryService = ServiceCategoryService.getInstance()
import '../mockservice/ServiceCategoryService.mock'

test("update score for one serviceCategory", () => {
    const updateScore = (item) => {
        serviceCategories[0] =  item.score + 1;

        let tree = testRenderer.toJSON()
        expect(tree).toMatchSnapshot()
    }

    const testRenderer = TestRenderer.create(
        <ServiceCategoryPills 
            serviceCategories={serviceCategories}
            updateScore={updateScore}
        />
    )

    let tree = testRenderer.toJSON()
    expect(tree).toMatchSnapshot()

    const testInstance = testRenderer.root

    const updateScoreButtons =  testInstance.findAllByProps({
        className: 'nav-item'
    })

    updateScoreButtons[1].props.onClick()

})

test("render serviceCategroy in ServiceCategoryPills correctly", () => {
    const testRenderer = TestRenderer.create(
        <ServiceCategoryPills 
            serviceCategories={serviceCategories}
        />
    )

    let tree = testRenderer.toJSON()
    expect(tree).toMatchSnapshot()

    const testInstance = testRenderer.root

    const ServiceCategories = testInstance.findAllByProps({
        className: 'nav-item'
    })

    expect(ServiceCategories.length).toBe(8)
})

test("render serviceCategroy in ServiceCategoryPills from service", () => {
    serviceCategoryService
        .findAllServiceCategories()
        .then(serviceCategories => {
            const testRenderer = TestRenderer.create(
                <ServiceCategoryPills 
                    serviceCategories={serviceCategories}
                />
            )

            let tree = testRenderer.toJSON()
            expect(tree).toMatchSnapshot()

            const testInstance = testRenderer.root

            const ServiceCategories = testInstance.findAllByProps({
                className: 'nav-item'
            })

            expect(ServiceCategories.length).toBe(5)
    })
})


