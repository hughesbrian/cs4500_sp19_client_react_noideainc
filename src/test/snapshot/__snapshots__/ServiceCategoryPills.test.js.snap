// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render serviceCategroy in ServiceCategoryPills correctly 1`] = `
<ul
  className="nav nav-pills nav-fill wd-shadow wd-padding-20"
>
  <li
    className="nav-item"
  >
    <a
      className="nav-link btn-lg text-center"
      href="/services"
    >
      <i
        className="fa fa-home"
      />
      <br />
      Home
    </a>
  </li>
  <li
    className="nav-item"
    onClick={[Function]}
  >
    <a
      className="nav-link btn-lg text-center"
      href="/services"
    >
      <i
        className="fa fa-bookmark"
      />
      <br />
    </a>
  </li>
  <li
    className="nav-item"
    onClick={[Function]}
  >
    <a
      className="nav-link btn-lg text-center"
      href="/services"
    >
      <i
        className="fa fa-bookmark"
      />
      <br />
      Fitness Trainer
    </a>
  </li>
  <li
    className="nav-item"
    onClick={[Function]}
  >
    <a
      className="nav-link btn-lg text-center"
      href="/services"
    >
      <i
        className="fa fa-bookmark"
      />
      <br />
      House Painter
    </a>
  </li>
  <li
    className="nav-item"
    onClick={[Function]}
  >
    <a
      className="nav-link btn-lg text-center"
      href="/services"
    >
      <i
        className="fa fa-briefcase"
      />
      <br />
      Business
    </a>
  </li>
  <li
    className="nav-item"
    onClick={[Function]}
  >
    <a
      className="nav-link btn-lg text-center"
      href="/services"
    >
      <i
        className="fa fa-heart"
      />
      <br />
      Wellness
    </a>
  </li>
  <li
    className="nav-item"
    onClick={[Function]}
  >
    <a
      className="nav-link btn-lg text-center"
      href="/services"
    >
      <i
        className="fa fa-paw"
      />
      <br />
      Pets
    </a>
  </li>
  <li
    className="nav-item"
  >
    <a
      className="nav-link btn-lg text-center"
      href="/services"
    >
      <i
        className="fa fa-ellipsis-h"
      />
      <br />
      More
    </a>
  </li>
</ul>
`;

exports[`render serviceCategroy in ServiceCategoryPills from service 1`] = `
<ul
  className="nav nav-pills nav-fill wd-shadow wd-padding-20"
>
  <li
    className="nav-item"
  >
    <a
      className="nav-link btn-lg text-center"
      href="/services"
    >
      <i
        className="fa fa-home"
      />
      <br />
      Home
    </a>
  </li>
  <li
    className="nav-item"
    onClick={[Function]}
  >
    <a
      className="nav-link btn-lg text-center"
      href="/services"
    >
      <i
        className="fa fa-bookmark"
      />
      <br />
      Tutoring
    </a>
  </li>
  <li
    className="nav-item"
    onClick={[Function]}
  >
    <a
      className="nav-link btn-lg text-center"
      href="/services"
    >
      <i
        className="fa fa-bookmark"
      />
      <br />
      Fitness Trainer
    </a>
  </li>
  <li
    className="nav-item"
    onClick={[Function]}
  >
    <a
      className="nav-link btn-lg text-center"
      href="/services"
    >
      <i
        className="fa fa-bookmark"
      />
      <br />
      House Painter
    </a>
  </li>
  <li
    className="nav-item"
  >
    <a
      className="nav-link btn-lg text-center"
      href="/services"
    >
      <i
        className="fa fa-ellipsis-h"
      />
      <br />
      More
    </a>
  </li>
</ul>
`;

exports[`update score for one serviceCategory 1`] = `
<ul
  className="nav nav-pills nav-fill wd-shadow wd-padding-20"
>
  <li
    className="nav-item"
  >
    <a
      className="nav-link btn-lg text-center"
      href="/services"
    >
      <i
        className="fa fa-home"
      />
      <br />
      Home
    </a>
  </li>
  <li
    className="nav-item"
    onClick={[Function]}
  >
    <a
      className="nav-link btn-lg text-center"
      href="/services"
    >
      <i
        className="fa fa-bookmark"
      />
      <br />
      Tutoring
    </a>
  </li>
  <li
    className="nav-item"
    onClick={[Function]}
  >
    <a
      className="nav-link btn-lg text-center"
      href="/services"
    >
      <i
        className="fa fa-bookmark"
      />
      <br />
      Fitness Trainer
    </a>
  </li>
  <li
    className="nav-item"
    onClick={[Function]}
  >
    <a
      className="nav-link btn-lg text-center"
      href="/services"
    >
      <i
        className="fa fa-bookmark"
      />
      <br />
      House Painter
    </a>
  </li>
  <li
    className="nav-item"
    onClick={[Function]}
  >
    <a
      className="nav-link btn-lg text-center"
      href="/services"
    >
      <i
        className="fa fa-briefcase"
      />
      <br />
      Business
    </a>
  </li>
  <li
    className="nav-item"
    onClick={[Function]}
  >
    <a
      className="nav-link btn-lg text-center"
      href="/services"
    >
      <i
        className="fa fa-heart"
      />
      <br />
      Wellness
    </a>
  </li>
  <li
    className="nav-item"
    onClick={[Function]}
  >
    <a
      className="nav-link btn-lg text-center"
      href="/services"
    >
      <i
        className="fa fa-paw"
      />
      <br />
      Pets
    </a>
  </li>
  <li
    className="nav-item"
  >
    <a
      className="nav-link btn-lg text-center"
      href="/services"
    >
      <i
        className="fa fa-ellipsis-h"
      />
      <br />
      More
    </a>
  </li>
</ul>
`;

exports[`update score for one serviceCategory 2`] = `
<ul
  className="nav nav-pills nav-fill wd-shadow wd-padding-20"
>
  <li
    className="nav-item"
  >
    <a
      className="nav-link btn-lg text-center"
      href="/services"
    >
      <i
        className="fa fa-home"
      />
      <br />
      Home
    </a>
  </li>
  <li
    className="nav-item"
    onClick={[Function]}
  >
    <a
      className="nav-link btn-lg text-center"
      href="/services"
    >
      <i
        className="fa fa-bookmark"
      />
      <br />
      Tutoring
    </a>
  </li>
  <li
    className="nav-item"
    onClick={[Function]}
  >
    <a
      className="nav-link btn-lg text-center"
      href="/services"
    >
      <i
        className="fa fa-bookmark"
      />
      <br />
      Fitness Trainer
    </a>
  </li>
  <li
    className="nav-item"
    onClick={[Function]}
  >
    <a
      className="nav-link btn-lg text-center"
      href="/services"
    >
      <i
        className="fa fa-bookmark"
      />
      <br />
      House Painter
    </a>
  </li>
  <li
    className="nav-item"
    onClick={[Function]}
  >
    <a
      className="nav-link btn-lg text-center"
      href="/services"
    >
      <i
        className="fa fa-briefcase"
      />
      <br />
      Business
    </a>
  </li>
  <li
    className="nav-item"
    onClick={[Function]}
  >
    <a
      className="nav-link btn-lg text-center"
      href="/services"
    >
      <i
        className="fa fa-heart"
      />
      <br />
      Wellness
    </a>
  </li>
  <li
    className="nav-item"
    onClick={[Function]}
  >
    <a
      className="nav-link btn-lg text-center"
      href="/services"
    >
      <i
        className="fa fa-paw"
      />
      <br />
      Pets
    </a>
  </li>
  <li
    className="nav-item"
  >
    <a
      className="nav-link btn-lg text-center"
      href="/services"
    >
      <i
        className="fa fa-ellipsis-h"
      />
      <br />
      More
    </a>
  </li>
</ul>
`;
